clang/runtime/dpct-rt/include/dpct/math.hpp

```diff
@@ -2216,6 +2218,94 @@ void ldmatrix(uintptr_t addr, T *m1, T *m2, T *m3, T *m4, bool trans = false) {
   ldmatrix(addr, m4, trans, 3);
 }
 
+/// Multiplies 2 matrices (A & B) and adds the result to C matrix and
+/// accumulates the result to a D matrix (MAD). Requires the sub-group size of
+/// kernel calling this function to be 32.
+/// \tparam [in] M The rows of A, C & D matrix
+/// \tparam [in] N The columns of B, C, D matrix
+/// \tparam [in] K The columns & rows of A & B matrices respectively
+/// \tparam [in] MulType The type used to multiply A and B matrix elements as
+/// \tparam [in] ABType The type of the input matrix (A & B) elements
+/// \tparam [in] CDType The type of the output matrix (C & D) elements
+/// \param [in] d The elements of the output D matrix to store the result to
+/// \param [in] a The elements of the input A matrix to be multiplied with B
+/// matrix elements
+/// \param [in] b The elements of the input B matrix to be multiplied with A
+/// matrix elements
+/// \param [in] c The elements of the input C matrix to be added with the result
+/// of A * B
+template <int M, int N, int K, typename MulType, typename ABType,
+          typename CDType, typename Op = sycl::bit_and<>>
+void mma(CDType **d, ABType *a, ABType *b, CDType *c, Op op = Op{}) {
+  auto sg = sycl::ext::oneapi::this_work_item::get_sub_group();
```

Comment:
I think pass an nd_item by parameter is better.

Reply:
Retained free-function style